### PR TITLE
Timesformer reader align

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
1. add TimeSformer's performance in README.md and README_cn.md
2. add TimeSformer's config file in `configs/recognition/timesformer/timesformer_k400_videos.yaml`
3. add TimeSformer's en doc and Zh-CN doc
4. add TimeSformer's two image about main arch and attention arch
5. add **UniformCrop** in `paddlevideo/loader/pipelines/augmentations.py`, and corresponding **UniformCropMetric** in `paddlevideo/metrics/uniform_crop_metric.py`
6. add pyav's video decode method in `paddlevideo/loader/pipelines/decode.py`
7. add linspace_sample strategy in `paddlevideo/loader/pipelines/sample.py`
8. add **VisionTransformer** in `paddlevideo/modeling/backbones/vit.py`
9. update the different forward method in `paddlevideo/modeling/framework/recognizers/recognizer2d.py ` while using TimeSformer and ViT based model
10. delete **coloring** and **AverageMeter** two unused import statement in `paddlevideo/tasks/train.py`
11. add an new training strategy **gradient accumulation** for TimeSformer and follow-up models, if used, in `paddlevideo/tasks/train.py`
12. fixed a bug when record the learning_rate in training process, when batch_size=1.
13. update the load_ckpt function for TimeSformer and follow-up ViT based models
14. update run.sh for TimeSformer
15. add a new InputSpec in `tools/export_model.py` for TimeSformer, even though there still remain some todo for inference model
16. add **TimeSformer_Inference_helper** class in `tools/utils.py `
17. add inference for timesformer and corresponding content in docs
18. fixed some hyper-link bugs in start.md
19. Update the description of `TenCrop`.
20. Change the order of data augmentation to align reader.
21. Add a new style of augmentation(almost in tensor mode), when backend is `pyav`
22. update `utils.summary` to fix some bugs where FLOPs could be negtive
23. Add trunc_normal_ function
24. Fix head/base.py line 108's bug

### CODE REVIEW

File: paddlevideo/loader/pipelines/decode.py [hunk 1]
```diff
@@ -53,11 +80,94 @@ def __call__(self, results):
...
+            num_clips = 1  # always be 1
+
+            # decode process
+            fps = float(container.streams.video[0].average_rate)
+
+            orig_duration = None  # None
+            tb = float(container.streams.video[0].time_base)
+            frames_length = container.streams.video[0].frames
+            duration = container.streams.video[0].duration
+            if duration is None and orig_duration is not None:
```
Comment (huangjun12) [inline paddlevideo/loader/pipelines/decode.py]: orig_duration would always be None?  这句条件一直是False，进不去？

File: paddlevideo/loader/pipelines/decode.py [hunk 2]
```diff
@@ -53,11 +80,94 @@ def __call__(self, results):
...
+                # If failed to fetch the decoding information, decode the entire video.
+                decode_all_video = True
+                video_start_pts, video_end_pts = 0, math.inf
+            else:
+                decode_all_video = False
+                start_idx, end_idx = get_start_end_idx(
+                    frames_length,
+                    self.sampling_rate * self.num_seg / self.target_fps * fps,
+                    clip_idx, num_clips)
+                timebase = duration / frames_length
```
Comment (huangjun12) [inline paddlevideo/loader/pipelines/decode.py:118]: Is the timebase the same as tb from float(container.streams.video[0].time_base) ?

File: paddlevideo/loader/pipelines/decode.py [hunk 3]
```diff
@@ -53,11 +80,94 @@ def __call__(self, results):
...
+
+                frames = [frame.to_rgb().to_ndarray() for frame in video_frames]
+                clip_sz = self.sampling_rate * self.num_seg / self.target_fps * fps
+
+                start_idx, end_idx = get_start_end_idx(
+                    len(frames),  # frame_len
+                    clip_sz,
+                    clip_idx
+                    if decode_all_video else 0,  #  -1 in train, 0 in test
+                    num_clips
```
Comment (huangjun12) [inline paddlevideo/loader/pipelines/decode.py]: num_clips if decode_all_video else 1   -->  num_clips  ?

as: num_clips always be 1

File: paddlevideo/loader/pipelines/decode.py [hunk 4]
```diff
@@ -53,11 +80,94 @@ def __call__(self, results):
...
+                container.close()
+
+                frames = [frame.to_rgb().to_ndarray() for frame in video_frames]
+                clip_sz = self.sampling_rate * self.num_seg / self.target_fps * fps
+
+                start_idx, end_idx = get_start_end_idx(
+                    len(frames),  # frame_len
+                    clip_sz,
+                    clip_idx
+                    if decode_all_video else 0,  #  -1 in train, 0 in test
```
Comment (huangjun12) [inline paddlevideo/loader/pipelines/decode.py]: 注释完善下: 
 If decode all video, -1 in train and valid, 0 in test; 
 else, always 0 in train, valid and test, as we has selected clip size frames when decode.

File: docs/zh-CN/model_zoo/recognition/timesformer.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+
+  当测试配置采用如下参数时，在Kinetics-400的validation数据集上的测试指标如下：
+
+   |      backbone      | Sampling method | num_seg | target_size | Top-1 |                         checkpoints                          |
+   | :----------------: | :-------------: | :-----: | :---------: | :---- | :----------------------------------------------------------: |
+   | Vision Transformer |   UniformCrop   |   8    |     224     | 77.25 | [TimeSformer_k400.pdparams](https://videotag.bj.bcebos.com/PaddleVideo-release2.2/TimeSformer_k400.pdparams) |
+
+
+- 测试时，TimeSformer视频采样策略为使用Linspace采样：时序上，从待采样的视频序列中均匀生成`num_seg`个稀疏采样点；空间上，选择长边两端及中间位置（左中右 或 上中下）3个区域采样。1个视频共采样1个clip。
```
Comment (huangjun12) [inline docs/zh-CN/model_zoo/recognition/timesformer.md]: 时序上，从待采样的视频序列中

从第几帧开始(0?) 到第几帧，均匀生成`num_seg`个稀疏采样点

还是从第0帧开始，以多少帧为间隔(smapling_rate=32?)，均匀生成`num_seg`个稀疏采样点

File: paddlevideo/loader/pipelines/decode.py (hunk 1 above)
Comment (HydrogenSulfate) [reply paddlevideo/loader/pipelines/decode.py]: 是的

File: paddlevideo/utils/save_load.py [hunk 6]
```diff
@@ -42,17 +46,75 @@ def load_ckpt(model, weight_path):
...
-        for item in tqdm(model.state_dict(), total=total_len, position=0):
-            name = item
-            desc.set_description('Loading %s' % name)
-            if state_dicts.get(name) is None and state_dicts.get('backbone.' + name) is not None:
-                tmp[name] = state_dicts['backbone.' + name]
-            else:
-                tmp[name] = state_dicts[name]
-            time.sleep(0.01)
+        if "VisionTransformer" in str(model):  # For TimeSformer case
+            tmp = state_dicts
```
Comment (huangjun12) [inline paddlevideo/utils/save_load.py]: For 2DCNN case: 
tmp = {}   //1. 新建空字典
state_dict = paddle.load(xxx)  // 2. load参数文件
tmp[name] = state_dicts[xxx]  //3. 使用参数文件，往空字典添加参数（包含参数转换操作）
model.set_state_dict(tmp)  // 4. 模型载入参数

能参考2DCNN的case重构下TimeSfomer预训练模型的载入逻辑吗？
其中参数转换的操作，可以在这个文件里，再单独封一个函数。
def pretrain_vit_param_trans():
   xxx

File: paddlevideo/modeling/backbones/vit.py [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+
+        x = self.norm(x)
+        return x[:, 0]  # [B, 1, embed_dim]
+
+    def forward(self, x):
+        x = self.forward_features(x)
+        return x
+
+
+if __name__ == '__main__':
```
Comment (huangjun12) [inline paddlevideo/modeling/backbones/vit.py]: 这里删一下？

File: paddlevideo/loader/pipelines/decode.py (hunk 1 above)
Comment (HydrogenSulfate) [reply paddlevideo/loader/pipelines/decode.py]: 已删除

File: paddlevideo/loader/pipelines/decode.py (hunk 2 above)
Comment (HydrogenSulfate) [reply paddlevideo/loader/pipelines/decode.py:118]: 已删除

File: paddlevideo/modeling/backbones/vit.py (hunk 7 above)
Comment (HydrogenSulfate) [reply paddlevideo/modeling/backbones/vit.py]: 已删除

File: paddlevideo/loader/pipelines/decode.py (hunk 2 above)
Comment (huangjun12) [reply paddlevideo/loader/pipelines/decode.py:118]: 这个check是一样的吗

File: paddlevideo/utils/save_load.py (hunk 6 above)
Comment (HydrogenSulfate) [reply paddlevideo/utils/save_load.py]: 已重构

File: docs/zh-CN/model_zoo/recognition/timesformer.md (hunk 5 above)
Comment (HydrogenSulfate) [reply docs/zh-CN/model_zoo/recognition/timesformer.md]: 已修改

File: paddlevideo/loader/pipelines/decode.py (hunk 4 above)
Comment (HydrogenSulfate) [reply paddlevideo/loader/pipelines/decode.py]: 已修改

File: paddlevideo/utils/save_load.py [hunk 8]
```diff
@@ -18,15 +18,78 @@
...
+    ret_str = "loading {:<20d} weights completed.".format(len(model.state_dict()))
+    desc.set_description(ret_str)
+    return new_state_dicts
 
 #XXX(shipping): maybe need load N times because of different cards have different params.
 @main_only
-def load_ckpt(model, weight_path):
+def load_ckpt(model,
+              weight_path,
+              num_patches=None,
```
Comment (huangjun12) [inline paddlevideo/utils/save_load.py]: 后面这三个参数用  **kargs，load接口里出现这三个参数有点奇怪

File: tools/summary.py [hunk 9]
```diff
@@ -67,12 +69,11 @@ def main():
...
     num_seg = args.num_seg
     #NOTE: only support tsm now, will refine soon
-    params_info = paddle.summary(model, (1, num_seg, 3, img_size, img_size))
+    params_info = paddle.summary(model, (1, 1, num_seg, 3, img_size, img_size))
     print(params_info)
 
     if args.FLOPs:
-        flops_info = paddle.flops(model, [1, num_seg, 3, img_size, img_size],
-                                  print_detail=True)
+        flops_info = paddleslim.analysis.flops(model, [1, 1, num_seg, 3, img_size, img_size])
```
Comment (huangjun12) [inline tools/summary.py:76]: paddle.flops接口跑不通吗？
